### PR TITLE
[CALCITE-5798] Improve simplification of '(x < y) IS NOT TRUE' when x and y are not nullable

### PR DESCRIPTION
**Pull request by herunkang2018**

### CODE REVIEW

File: core/src/test/java/org/apache/calcite/rex/RexProgramTest.java [hunk 1]
```diff
@@ -2010,6 +2010,20 @@ private void checkExponentialCnf(int n) {
...
+  /** Unit test for
+   * <a href="https://issues.apache.org/jira/browse/CALCITE-5798">[CALCITE-5798]
+   * '(1 < x) IS NOT TRUE' when x is not nullable should be simplified to '1 >= x'</a>. */
+  @Test void testSimplifyInverseComparisonWithIsNotNullPredicate() {
+    final RexNode ref = input(tInt(true), 0);
+    RelOptPredicateList relOptPredicateList =
+        RelOptPredicateList.of(rexBuilder,
+            ImmutableList.of(isNotNull(ref)));
+    checkSimplifyFilter(
+        isNotTrue(gt(literal(1), ref)),
```
Comment (NobiGo) [inline core/src/test/java/org/apache/calcite/rex/RexProgramTest.java:2022]: Looks like the test is the opposite. The summary is '(1 < x) IS NOT TRUE-->1>=x' but the test is '(1>x) IS NOT TRUE-->1<=x'. The test result is good. Just I think we should stay the same logic.

**Review by NobiGo — CHANGES_REQUESTED** [review]
Only need a small change. Others look good to me.

File: core/src/test/java/org/apache/calcite/rex/RexProgramTest.java (hunk 1 above)
Comment (herunkang2018) [reply core/src/test/java/org/apache/calcite/rex/RexProgramTest.java:2022]: @NobiGo Thanks for review, I will refine it soon.

File: core/src/main/java/org/apache/calcite/plan/RelOptPredicateList.java [hunk 2]
```diff
@@ -236,14 +236,18 @@ public boolean isEffectivelyNotNull(RexNode e) {
...
       List<RexNode> operands = ((RexCall) e).getOperands();
       // We can have just one operand in case e.g. of a RexSubQuery with IN operator.
       if (operands.size() > 1 && operands.get(1) instanceof RexLiteral
           && !((RexLiteral) operands.get(1)).isNull()) {
         return isEffectivelyNotNull(operands.get(0));
       }
+      if (operands.size() > 1 && operands.get(0) instanceof RexLiteral
+          && !((RexLiteral) operands.get(0)).isNull()) {
+        return isEffectivelyNotNull(operands.get(1));
+      }
```
[the remark this answers is not in the document]
Comment (herunkang2018) [reply core/src/main/java/org/apache/calcite/plan/RelOptPredicateList.java:250]: @kgyrtkirk Thanks for your advice. The refactor logic is more genaral, to covers more case like `x < y is not true`. I will refactor it soon.

[the remark this answers is not in the document]
Comment (herunkang2018) [reply core/src/main/java/org/apache/calcite/plan/RelOptPredicateList.java:250]: For the first advice, the RexLiteral doesn't need to process seperately, it is covered by L229.

**Comment by herunkang2018** [comment]
After refactor with @kgyrtkirk's advice, I changed the title to 'Improve simplification of '(x < y) IS NOT TRUE' and '(x < y) IS TRUE' when x and y is not nullable.'

File: core/src/test/java/org/apache/calcite/rex/RexProgramTest.java [hunk 3]
```diff
@@ -2010,6 +2010,57 @@ private void checkExponentialCnf(int n) {
...
+        RexUnknownAs.UNKNOWN,
+        "<($0, $1)");
+
+    // "(x < y) IS NOT TRUE" (if x and y are both not nullable) => "x >= y"
+    checkSimplifyFilter(
+        isNotTrue(lt(xRef, yRef)),
+        relOptPredicateList,
+        ">=($0, $1)");
+
+    // "(x < 1) IS TRUE" (if x is not nullable) => "x < 1"
```
Comment (herunkang2018) [inline core/src/test/java/org/apache/calcite/rex/RexProgramTest.java:2037]: These two test cases above are not related to this issue, they can be passed before. I add them for better code coverage.

**Review by NobiGo — APPROVED** [review]
LGTM

**Comment by herunkang2018** [comment]
@NobiGo @kgyrtkirk Thanks a lot for reviewing. 
I have changed the PR's title and Jira's title to "Improve simplification of '(x < y) IS NOT TRUE' when x and y are not nullable" to make it more clear.

**Comment by NobiGo** [comment]
Please rebase the commits. I will merge it.

**Comment by herunkang2018** [comment]
@NobiGo I have rebased. Please have a look, thanks.

**Pull request merged**